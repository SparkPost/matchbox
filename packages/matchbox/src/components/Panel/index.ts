export { default as Panel } from './Panel';
export * from './Panel';
export * from './Section';
export * from './Header';
export * from './SubHeader';
export * from './Action';
