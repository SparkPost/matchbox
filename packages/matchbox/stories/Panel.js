import React, { Fragment } from 'react';
import { storiesOf } from '@storybook/react';
import { withInfo } from '@storybook/addon-info';
import { action } from '@storybook/addon-actions';
import { StoryContainer } from './helpers';

import { Panel, Action } from '../src';

storiesOf('Panel', module)
  .addDecorator((getStory) => (
    <StoryContainer>{ getStory() }</StoryContainer>
  ))
  .add('with a title', withInfo()(() => (
    <Panel title="Title" sectioned>This is a panel with a title</Panel>
  )))

  .add('with an accent', withInfo()(() => (
    <Fragment>
      <Panel accent sectioned>This is a highlighted panel with a title</Panel>
      <Panel accent='navy' sectioned>This is a highlighted panel with a title</Panel>
    </Fragment>
  )))

  .add('with a footer', withInfo()(() => (
    <Fragment>
      <Panel title="Title" accent sectioned>This is a panel</Panel>
      <Panel.Footer left='Left aligned' right='Right aligned'></Panel.Footer>
    </Fragment>
  )))

  .add('with multiple sections',
  withInfo()(() => (
      <Panel>
        <Panel.Section>This is a panel with sections</Panel.Section>
        <Panel.Section>This is a panel with sections</Panel.Section>
        <Panel.Section>This is a panel with sections</Panel.Section>
      </Panel>
  )))

  .add('with actions', withInfo()(() => {
    const actions = [
      {
        content: 'Edit',
        onClick: action('Edit Clicked'),
        color: 'red'
      },
      {
        content: 'Delete',
        onClick: action('Delete Clicked'),
        color: 'red'
      },
    ];
    const sectionActions = [
      {
        content: 'View Details',
        onClick: action('Details Clicked'),
        color: 'red'
      }
    ];
    return (
        <Panel
          actions={actions}
          accent='red'
          title='Panel with Actions'>
          <Panel.Section
            actions={sectionActions} >
            <p>Lorem ipsum dolor sit amet, consectetur adipisicing elit. Amet perspiciatis harum reprehenderit, odio temporibus culpa beatae iure!</p>
          </Panel.Section>
        </Panel>
    );
  }));
