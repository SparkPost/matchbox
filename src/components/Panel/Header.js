import React, { Component } from 'react';
import { buttonsFrom } from '../Button';
import styles from './Panel.module.scss';

const actionOverrides = { plain: true, size: 'small' };

const Header = ({ title, actions }) => {
  const actionMarkup = actions && actions.length
        ? <div className={styles.Actions}>{ buttonsFrom(actions, actionOverrides) }</div>
        : null;

  return (
    <div className={styles.Header}>
      { actionMarkup }
      <h2 className={styles.HeaderText}>{ title }</h2>
    </div>
  );
};

export default Header;
