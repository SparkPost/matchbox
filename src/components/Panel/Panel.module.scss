@import '../../styles/config';

.Panel {
  position: relative;
  margin-bottom: rem(18);
  background: #fff;
  box-shadow: shadow();
  border-radius: border-radius();
}

.accent {
  &:before {
    display: block;
    content: "";
    position: absolute;
    top: 0;
    left: 0;
    right: 0;
    background: color(orange);
    height: 3px;
    border-radius: border-radius() border-radius() 0 0;
  }
}

.Header {
  padding: rem(18);
  border-bottom: 1px solid color(gray, 9);
  font-weight: 600;
}

.HeaderText {
  margin: 0;
}

.Body {
  padding: rem(18);
  border-bottom: 1px solid color(gray, 9);

  &:last-child {
    border-bottom: none;
  }
}
