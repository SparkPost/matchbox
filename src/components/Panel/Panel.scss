.Panel {
  position: relative;
  background: #fff;
  box-shadow: 0px 1px 3px #E1E1E6;
  border-radius: 2px;
}

.Panel-accent {
  &:before {
    display: block;
    content: "";
    position: absolute;
    top: 0;
    left: 0;
    right: 0;
    background: adjust-hue(orange, -10);
    height: 3px;
    border-radius: 2px 2px 0 0;
  }
}

.Header {
  padding: 21px;
  border-bottom: 1px solid #E1E1E6;
  font-weight: 600;

  & > * {
    margin: 0;
  }
}

.Body {
  padding: 21px;
}
