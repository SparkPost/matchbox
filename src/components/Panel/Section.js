import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { buttonsFrom } from '../Button';
import styles from './Panel.module.scss';

const actionOverrides = { plain: true, size: 'small' };

class Section extends Component {
  static propTypes = {
     /**
      * Actions that build buttons. Most button props will work in here.
      * e.g. { content: 'button label', onClick: callback() }
      */
     actions: PropTypes.arrayOf(PropTypes.shape({
       content: PropTypes.string.isRequired
     })),
     /**
      * Panel Content
      */
     children: PropTypes.oneOfType([
       PropTypes.arrayOf(PropTypes.node),
       PropTypes.node
     ]),
  };

  render() {
    const {
      children,
      actions
    } = this.props;

    const actionMarkup = actions && actions.length
          ? <div className={styles.Actions}>{ buttonsFrom(actions, actionOverrides) }</div>
          : null;

    return (
      <div className={styles.Body}>
        <div className={styles.SectionContent}>
          { children }
        </div>
        { actionMarkup }
      </div>
    );
  }
}

Section.displayName = 'Panel.Section';
export default Section;
