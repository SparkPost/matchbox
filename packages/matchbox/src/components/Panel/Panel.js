import React, { Component } from 'react';
import PropTypes from 'prop-types';
import classnames from 'classnames';

import Section from './Section';
import Footer from './Footer';
import Header from './Header';
import styles from './Panel.module.scss';

class Panel extends Component {
  static Section = Section;
  static Footer = Footer;

  static displayName = 'Panel';

  static propTypes = {
    /**
     * The panel heading title
     */
    title: PropTypes.node,
    /**
      * Shows an accent bar, defaults to orange if boolean
      */
    accent: PropTypes.oneOfType([
      PropTypes.bool,
      PropTypes.oneOf(['orange', 'blue', 'navy', 'purple', 'red', 'yellow', 'green', 'gray'])
    ]),
    /**
      * Adds a padded section automatically
      */
    sectioned: PropTypes.bool,
    /**
      * Actions that build buttons. Most button props will work in here.
      * e.g. { content: 'button label', onClick: callback() }
      */
    actions: PropTypes.arrayOf(PropTypes.shape({
      content: PropTypes.node.isRequired
    })),
    /**
      * Panel Content
      */
    children: PropTypes.node
  };

  render() {
    const {
      children,
      title,
      accent,
      sectioned,
      actions,
      className,
      ...rest
    } = this.props;

    const accentColor = accent === true ? 'orange' : accent;

    const headerMarkup = title
      ? <Header title={title} actions={actions} />
      : null;

    const contentMarkup = sectioned
      ? <Section>{children}</Section>
      : children;

    const accentMarkup = accentColor
      ? <div className={classnames(styles.Accent, styles[`accent-${accentColor}`])} />
      : null;

    const panelStyles = classnames(styles.Panel, className);

    return (
      <div className={panelStyles} {...rest}>
        {accentMarkup}
        {headerMarkup}
        {contentMarkup}
      </div>
    );
  }
}

export default Panel;
