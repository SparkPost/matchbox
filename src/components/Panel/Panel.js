import React, { Component } from 'react';
import PropTypes from 'prop-types';
import classnames from 'classnames';

import Section from './Section';
import Header from './Header';
import styles from './Panel.module.scss';

class Panel extends Component {
  static Section = Section;

  static propTypes = {
    /**
     * The panel heading title
     */
     title: PropTypes.string,
     /**
      * Shows an orange accent bar
      */
     accent: PropTypes.bool,
     /**
      * Adds a padded section automatically
      */
     sectioned: PropTypes.bool,
     /**
      * Actions that build buttons. Most button props will work in here.
      * e.g. { content: 'button label', onClick: callback() }
      */
     actions: PropTypes.arrayOf(PropTypes.shape({
       content: PropTypes.string.isRequired
     })),
     /**
      * Panel Content
      */
     children: PropTypes.oneOfType([
       PropTypes.arrayOf(PropTypes.node),
       PropTypes.node
     ]),
  };

  render() {
    const {
      children,
      title,
      accent,
      sectioned,
      actions,
      ...rest
    } = this.props;

    const headerMarkup = title
      ? <Header title={title} actions={actions} />
      : null;

    const contentMarkup = sectioned
      ? <Section>{children}</Section>
      : children;

    const panelStyles = classnames(styles.Panel, accent && styles.accent);

    return (
      <div className={panelStyles} {...rest}>
        { headerMarkup }
        { contentMarkup }
      </div>
    );
  }
};

export default Panel;
