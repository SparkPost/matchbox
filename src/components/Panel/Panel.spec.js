import React from 'react';
import renderer from 'react-test-renderer';

import Panel from './Panel';

describe('snapshot tests', () => {

  test('Panel will render correctly with all options', () => {
    expect(renderer.create(
      <Panel>Foo</Panel>
    )).toMatchSnapshot();
  });

});
