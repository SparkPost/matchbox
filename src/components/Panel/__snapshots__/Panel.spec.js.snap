// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`snapshot tests Panel will render correctly with all options 1`] = `
<div
  className=""
>
  Foo
</div>
`;
